README.md
# Blade Engine

This game engine and game itself is developed for our WIP title called Runblader.
We aim to replace the requirements for using Quake 2 data and introduce our own replacements
to become Quake 2 independent. Our interest is in Quake 2 mechanics and tech itself and Quake 2
related assets are not planned to be used within our title.

This code is build upon Blade Engine, which itself is based on Icculus Quake II. 
Blade Engine is released under the terms of the GPL version 2. See the
LICENSE file for further information.


## Installation

Blade Engine is installed in 2 steps:

- Game data setup.
- Download and extract the executables.


### Game data Setup

There is currently no game data available at the moment.

Quake 2 game data and/or associated mods are not supported!

### Download and extract the executables

How the Blade Engine executables are installed depends on the platform:

- For Windows a package with all Blade Engine executables *will be* provided.
- There is no Linux or *BSD package available so far;
- On OS X you need to compile from source.
- Of course Blade Engine can be compiled from source on all platforms.


#### Microsoft Windows

1. Get the latest release from your game archive.
2. Extract it into the gamedata directory created above. blade.exe must be
   placed next to the baseb/ subdirectory.

On Windows Blade Engine is fully portable, you can move the installation
directory wherever and whenever you like. To update Blade Engine just
overwrite the old files with the new ones.


#### Binary Package from your Linux distribution or BSD system

TBD

### Compiling from source

To compile Blade Engine from source, you need the following dependencies,
including development headers:

- A GCC-compatible compiler like GCC, MinGW (see below) or clang.
- GNU make.
- A libGL implementation with OpenGL system headers.
- SDL 2.0 or SDL 1.2 (2.0 recommended, edit the Makefile to use 1.2 instead).
- libogg, libvorbis and libvorbisfile.
- A OpenAL implementation, openal-soft is highly recommended.
- zlib.

#### On Linux distribution or BSD systems

On debian based distributions (including Ubuntu and Mint) the dependencies can
be installed with: `apt-get install build-essential libgl1-mesa-dev libsdl2-dev
libogg-dev libvorbis-dev libopenal-dev zlib1g-dev`

On OS X we recommend using [homebrew](https://brew.sh) to install the
dependencies: `brew install sdl2 libvorbis libogg openal-soft`

On FreeBSD you'll need something like: `pkg install gmake libGL sdl2 libogg
libvorbis openal-soft`


#### On Windows

On Windows a MinGW environment is needed. A preinstalled environment with all
dependencies can be found at https://deponie.yamagi.org/quake2/windows/build/
Just extract it into C:\MSYS2\ and start either the 32 bit or 64 bit version
through *C:\MSYS2\msys32.exe* or *C:\MSYS2\msys64.exe*. With the preinstalled
MinGW environment GNU Make is highly recommended.


#### Compiling

Change into the code/ source directory and type *make*. 
After that copy everything from the release/
directory to your Blade Engine installation directory.


CHANGELOG
Blade 0.1a:
- New fork of a source port with new features and changes applied.
- Got rid of software renderer as it became redundant.
- New main menu code changes and animated background support.
- Rework of some engine parts to better acommodate our dev workflow.
- Increment game api version.
- Add provisionary gun reload mechanic.
- Small code organisation changes.

Quake II 7.10 to 7.20:
- Add the soft renderer back. This feature is considered experimental.
  The porting of the old soft renderer code to SDL and it's cleanup
  were done by Denis Pauk.
- Rename several cvars to be consistent across different renderers.
  The configuration file will be converted at the first start, when an
  old cvar name is used a message is printed.
- Make the client unicode compatible. Yamagi Quake II can now be 
  installed into pathes with unicode characters in it. On Windows the
  user name of the current account may contain unicode characters.
  As a side effect the game can run on ReFS volumes. While '+set
  basedir' is still supported, the unicode compatible '-datadir'
  should be used.
- Another round of timing fixes. The game is now capable of holding
  the requestes or vsync framerate even on slow machines and with
  problematic GL drivers.
- Fix server side mod handling, their configs are now saved to the
  correct directories and the configs are reexeced at mod startup.

Quake II 7.02 to 7.10:
- Joystick support including haptic feedback. This fantastic work was
  done by Denis Pauk. The dirty work is done by SDL, how good or bad
  a joystick or gamepad is supported depends on SDLs support for it.
- Fix the old SDL sound backend, s_openal set to 0 is working again.
- Fix possible Vorbis buffer underruns if too many sound samples are
  in flight. This occured only in large multi player games with at
  least 6 custom models.
- Fix a possible crash on Windows if MSAA was set to a value not
  supported by the driver.
- It's now possible to play through the whole game on a Raspberry PI
  and other ARM boards. Please note that the RPIs hardware is really
  limited. Only the OpenGL 1.4 renderer is supported and the framerate
  is highly dependend on the screen resolution.

Quake II 7.01 to 7.02:
- Fix several corner cases regarding render library loading. The game
  should now always fall back to the OpenGL 1.4 renderer if the new
  OpenGL 3.2 renderer can't be initialized. Also the game aborts if no
  useable OpenGL implementation exists.
- Refactor the search path code. This should fix several bugs with
  Quake II writing to the wrong directories or being unable to find
  some / all assets.
- Reimplement portable binaries. If called with the -portable command
  line option Quake II saves all data (configs, savegames, screenshorts
  etc.) into it's systemwide installation directory and not users home
  directory. In contrast to the old implementation on Windows stdout.txt
  contains all output, the first lines are no longer missing.
- vid_fullscreen set to 1 now keeps the desktops resolution. Set it to 2
  to change the resolution.
- Instead of a list with precalculated FOV values the video menu now
  shows a slider with possible values from 60 to 120. Horplus is now
  always enabled, set the horplus cvar to 0 to disable it.
- The game is now able to hold the requested framerate (either by the
  vsync or the gl_maxfps cvar) with an accuracy of about +/- 1% as long
  as the hardware is fast enough.  The framecounter was reimplemented
  to be much more precise.
- Fix misspredictions if an original client running on Win32 connects
  to a Yamagi Quake II server running on Linux/i386.

Quake II 7.00 to 7.01:
- Fix build of GL3 for platforms without SSE.
- Fix Jennel Jaquays name in credits and quit screen.
- Make Quake II high DPI aware on Window Vista and above.
- Fix some problems with loading dependend librarys on Windows.

Quake II 6.00 to 7.00:
- Remove the broken multitexturing render path from the OpenGL 1.4
  renderer. It was switched off by default in 6.00.
- Reimplement the support for shared renderer libraries. Please note
  the this is an incompatible implementation with an custom API. The
  original renderer libraries will not work!
- Implement an OpenGL 3.2 renderer. This renderer has the same look
  and feel as the old OpenGL 1.4 renderer but makes heavy use of
  modern OpenGL and GPU features. An OpenGL 3.2 capable GPU (Intel
  starting with Ivy Bridge on Windows or Sandy Bridge on Linux, Nvidia
  staring with G80 and AMD starting with R600 / HD2000) is required.
- Fix OpenAL compatibility with modern openal-soft versions.
- Several fixes and optimizations to OpenAL, implement support for
  doppler effects. (by xorw)

Quake II 5.34 to 6.00:
- Make the client asynchronous. The old behaviour can be forced by
  setting cl_async to 0. Please note that asynchronicity can lead to
  problems if the old SDL 1.2 backend is used and vsync is enabled.
- Implement gl_overbrightbits in the non multitexturing case.  A value
  of 1 just fixes lighting on water surfaces, higher values increase
  the brightness of everything.
- General renderer overhaul for better compatibility with modern GPUs.
  OpenGL 1.4 is now required, older versions are no longer supported.
  Multitexturing was deprecated and will be removed in a future release.
- Fix some longstanding AI problems.
- Several general gameplay fixes.

Quake II 5.33 to 5.34:
- Add support for stereo 3D (by Valery Guskov)
- Make gibt solid so they move with conveyor belts.
- Disable gl_ext_multitexturing by default.
- Switch from an arch whitelist to an "all archs are supported"
  approach.
- Add a new README.

Quake II 5.32 to 5.33:
- Add OGG volume slider in settings menu
- Fixed some bugs in volume settings
- Replaced HUD scale option in video menu with generic UI scale
- General UI upscaling improvements
- Better support for keyboards with AZERTY layout

Quake II 5.31 to 5.32:
- Fix problems with never openal-soft versions.

src/common/frame.c
/*
 * Copyright (C) 1997-2001 Id Software, Inc.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
 * 02111-1307, USA.
 *
 * =======================================================================
 *
 * Platform independent initialization, main loop and frame handling.
 *
 * =======================================================================
 */

#include "header/common.h"
#include "header/zone.h"
#include <setjmp.h>

cvar_t *developer;
cvar_t *modder;
cvar_t *timescale;
cvar_t *fixedtime;
cvar_t *cl_maxfps;
cvar_t *dedicated;
cvar_t *busywait;

extern cvar_t *logfile_active;
extern jmp_buf abortframe; /* an ERR_DROP occured, exit the entire frame */
extern zhead_t z_chain;

#ifndef DEDICATED_ONLY
FILE *log_stats_file;
cvar_t *cl_async;
cvar_t *cl_timedemo;
cvar_t *vid_maxfps;
cvar_t *host_speeds;
cvar_t *log_stats;
cvar_t *showtrace;
#endif

// Forward declarations
#ifndef DEDICATED_ONLY
int GLimp_GetRefreshRate(void);
qboolean R_IsVSyncActive(void);
#endif

/* host_speeds times */
int time_before_game;
int time_after_game;
int time_before_ref;
int time_after_ref;

// Used in the network- and input pathes.
int curtime;

#ifndef DEDICATED_ONLY
void Key_Init(void);
void SCR_EndLoadingPlaque(void);
#endif

// Is the game portable?
qboolean is_portable;

// Game given by user
char userGivenGame[MAX_QPATH];

// ----

static void
Qcommon_Buildstring(void)
{
	int i;
	int verLen;
	const char* versionString;


	versionString = va("Blade Engine v%s", BDEVERSION);
	verLen = strlen(versionString);

	printf("\n%s\n", versionString);

	for( i = 0; i < verLen; ++i)
	{
		printf("=");
	}

	printf("\n");


#ifndef DEDICATED_ONLY
	printf("Client build options:\n");
#ifdef SDL2
	printf(" + SDL2\n");
#else
	printf(" - SDL2 (using 1.2)\n");
#endif

#ifdef CDA
	printf(" + CD audio\n");
#else
	printf(" - CD audio\n");
#endif
#ifdef OGG
	printf(" + OGG/Vorbis\n");
#else
	printf(" - OGG/Vorbis\n");
#endif
#ifdef USE_OPENAL
	printf(" + OpenAL audio\n");
#else
	printf(" - OpenAL audio\n");
#endif
#ifdef ZIP
	printf(" + Zip file support\n");
#else
	printf(" - Zip file support\n");
#endif
#endif

	printf("Platform: %s\n", BDEOSTYPE);
	printf("Architecture: %s\n", BDEARCH);
}

#ifndef DEDICATED_ONLY
#define FRAMEDELAY 5
#else
#define FRAMEDELAY 850
#endif

void
Qcommon_Mainloop(void)
{
	long long newtime;
	long long oldtime = Sys_Microseconds();

	/* The mainloop. The legend. */
	while (1)
	{
		// Throttle the game a little bit.
		if (busywait->value)
		{
			long long spintime = Sys_Microseconds();

			while (1)
			{
#if defined (__GNUC__) && (__i386 || __x86_64__)
				/* Give the CPU a hint that this is a very tight
				   spinloop. One PAUSE instruction each loop is
				   enough to reduce power consumption and head
				   dispersion a lot, it's 95°C against 67°C on
				   a Kaby Lake laptop. */
				asm("pause");
#endif

				if (Sys_Microseconds() - spintime >= FRAMEDELAY)
				{
					break;
				}
			}
		}
		else
		{
			Sys_Nanosleep(FRAMEDELAY * 1000);
		}

		newtime = Sys_Microseconds();
		Qcommon_Frame(newtime - oldtime);
		oldtime = newtime;
	}
}

void Qcommon_ExecConfigs(qboolean gameStartUp)
{
	Cbuf_AddText("exec default.cfg\n");
	Cbuf_AddText("exec BDE.cfg\n");
	Cbuf_AddText("exec config.cfg\n");
	if(gameStartUp)
	{
		// only when the game is first started we execute autoexec.cfg and set the cvars from commandline
		Cbuf_AddText("exec autoexec.cfg\n");
		Cbuf_AddEarlyCommands(true);
	}
	Cbuf_Execute();
}

void
Qcommon_Init(int argc, char **argv)
{
	// Jump point used in emergency situations.
	if (setjmp(abortframe))
	{
		Sys_Error("Error during initialization");
	}

	// Print the build and version string
	Qcommon_Buildstring();

	// Seed PRNG
	randk_seed();

	// Initialize zone malloc().
	z_chain.next = z_chain.prev = &z_chain;

	// Start early subsystems.
	COM_InitArgv(argc, argv);
	Swap_Init();
	Cbuf_Init();
	Cmd_Init();
	Cvar_Init();

#ifndef DEDICATED_ONLY
	Key_Init();
#endif

	/* we need to add the early commands twice, because
	   a basedir or cddir needs to be set before execing
	   config files, but we want other parms to override
	   the settings of the config files */
	Cbuf_AddEarlyCommands(false);
	Cbuf_Execute();

	// remember the initial game name that might have been set on commandline
	{
		cvar_t* gameCvar = Cvar_Get("game", "", CVAR_LATCH | CVAR_SERVERINFO);
		const char* game = "";

		if(gameCvar->string && gameCvar->string[0])
		{
			game = gameCvar->string;
		}

		Q_strlcpy(userGivenGame, game, sizeof(userGivenGame));
	}

	// The filesystems needs to be initialized after the cvars.
	FS_InitFilesystem();

	// Add and execute configuration files.
	Qcommon_ExecConfigs(true);

	// Zone malloc statistics.
	Cmd_AddCommand("z_stats", Z_Stats_f);

	// cvars

	cl_maxfps = Cvar_Get("cl_maxfps", "60", CVAR_ARCHIVE);

	developer = Cvar_Get("developer", "0", 0);
	fixedtime = Cvar_Get("fixedtime", "0", 0);

	logfile_active = Cvar_Get("logfile", "1", CVAR_ARCHIVE);
	modder = Cvar_Get("modder", "0", 0);
	timescale = Cvar_Get("timescale", "1", 0);

	char *s;
	s = va("%s %s %s %s", BDEVERSION, BDEARCH, BUILD_DATE, BDEOSTYPE);
	Cvar_Get("version", s, CVAR_SERVERINFO | CVAR_NOSET);
	busywait = Cvar_Get("busywait", "1", CVAR_ARCHIVE);

#ifndef DEDICATED_ONLY
	cl_async = Cvar_Get("cl_async", "1", CVAR_ARCHIVE);
	cl_timedemo = Cvar_Get("timedemo", "0", 0);
	dedicated = Cvar_Get("dedicated", "0", CVAR_NOSET);
	vid_maxfps = Cvar_Get("vid_maxfps", "95", CVAR_ARCHIVE);
	host_speeds = Cvar_Get("host_speeds", "0", 0);
	log_stats = Cvar_Get("log_stats", "0", 0);
	showtrace = Cvar_Get("showtrace", "0", 0);
#else
	dedicated = Cvar_Get("dedicated", "1", CVAR_NOSET);
#endif

	// We can't use the clients "quit" command when running dedicated.
	if (dedicated->value)
	{
		Cmd_AddCommand("quit", Com_Quit);
	}

	// Start late subsystem.
	Sys_Init();
	NET_Init();
	Netchan_Init();
	SV_Init();
#ifndef DEDICATED_ONLY
	CL_Init();
#endif

	// Everythings up, let's add + cmds from command line.
	if (!Cbuf_AddLateCommands())
	{
		if (!dedicated->value)
		{
			// Start main menu
			Cbuf_AddText("start_intro\n");
		}
		else
		{
			// ...or dedicated server.
			Cbuf_AddText("dedicated_start\n");
		}

		Cbuf_Execute();
	}
#ifndef DEDICATED_ONLY
	else
	{
		/* the user asked for something explicit
		   so drop the loading plaque */
		SCR_EndLoadingPlaque();
	}
#endif

	Com_Printf("==== Blade Engine Initialized ====\n\n");
	Com_Printf("*************************************\n\n");

	// Call the main loop
	Qcommon_Mainloop();
}

#ifndef DEDICATED_ONLY
void
Qcommon_Frame(int msec)
{
	// Used for the dedicated server console.
	char *s;

	// Statistics.
	int time_before = 0;
	int time_between = 0;
	int time_after;

	// Target packetframerate.
	int pfps;

	//Target renderframerate.
	int rfps;

	// Time since last packetframe in microsec.
	static int packetdelta = 1000000;

	// Time since last renderframe in microsec.
	static int renderdelta = 1000000;

	// Accumulated time since last client run.
	static int clienttimedelta = 0;

	// Accumulated time since last server run.
	static int servertimedelta = 0;

	/* A packetframe runs the server and the client,
	   but not the renderer. The minimal interval of
	   packetframes is about 10.000 microsec. If run
	   more often the movement prediction in pmove.c
	   breaks. That's the Q2 variant if the famous
	   125hz bug. */
	qboolean packetframe = true;

	/* A rendererframe runs the renderer, but not the
	   client. The minimal interval is about 1000
	   microseconds. */
	qboolean renderframe = true;

	// Average time needed to process a render frame.
	static int avgrenderframetime;
	static int renderframetimes[60];
	static qboolean last_was_renderframe;

	// Average time needed to process a packet frame.
	static int avgpacketframetime;
	static int packetframetimes[60];
	static qboolean last_was_packetframe;


	/* In case of ERR_DROP we're jumping here. Don't know
	   if that' really save but it seems to work. So leave
	   it alone. */
	if (setjmp(abortframe))
	{
		return;
	}


	if (log_stats->modified)
	{
		log_stats->modified = false;

		if (log_stats->value)
		{
			if (log_stats_file)
			{
				fclose(log_stats_file);
				log_stats_file = 0;
			}

			log_stats_file = Q_fopen("stats.log", "w");

			if (log_stats_file)
			{
				fprintf(log_stats_file, "entities,dlights,parts,frame time\n");
			}
		}
		else
		{
			if (log_stats_file)
			{
				fclose(log_stats_file);
				log_stats_file = 0;
			}
		}
	}


	// Timing debug crap. Just for historical reasons.
	if (fixedtime->value)
	{
		msec = (int)fixedtime->value;
	}
	else if (timescale->value)
	{
		msec *= timescale->value;
	}


	if (showtrace->value)
	{
		extern int c_traces, c_brush_traces;
		extern int c_pointcontents;

		Com_Printf("%4i traces  %4i points\n", c_traces, c_pointcontents);
		c_traces = 0;
		c_brush_traces = 0;
		c_pointcontents = 0;
	}

	// vid_maxfps > 1000 breaks things, and so does <= 0
	// so cap to 1000 and treat <= 0 as "as fast as possible", which is 1000
	if (vid_maxfps->value > 1000 || vid_maxfps->value < 1)
	{
		Cvar_SetValue("vid_maxfps", 1000);
	}

	if(cl_maxfps->value > 250)
	{
		Cvar_SetValue("cl_maxfps", 130);
	}
	else if(cl_maxfps->value < 1)
	{
		Cvar_SetValue("cl_maxfps", 60);
	}


	// Save global time for network- und input code.
	curtime = Sys_Milliseconds();


	// Calculate target packet- and renderframerate.
	if (R_IsVSyncActive())
	{
		rfps = GLimp_GetRefreshRate();

		if (rfps > vid_maxfps->value)
		{
			rfps = (int)vid_maxfps->value;
		}
	}
	else
	{
		rfps = (int)vid_maxfps->value;
	}

	/* The target render frame rate may be too high. The current
	   scene may be more complex than the previous one and SDL
	   may give us a 1 or 2 frames too low display refresh rate.
	   Add a security magin of 5%, e.g. 60fps * 0.95 = 57fps. */
	pfps = (cl_maxfps->value > rfps) ? floor(rfps * 0.95) : cl_maxfps->value;


	/* Calculate average time spend to process a render
	   frame. This is highly depended on the GPU and the
	   scenes complexity. Take the last 60 pure render
	   frames (frames that are only render frames and
	   nothing else) into account and add a security
	   margin of 2%. */
	if (last_was_renderframe && !last_was_packetframe)
	{
		int measuredframes = 0;
		static int renderframenum;

		renderframetimes[renderframenum] = msec;

		for (int i = 0; i < 60; i++)
		{
			if (renderframetimes[i] != 0)
			{
				avgrenderframetime += renderframetimes[i];
				measuredframes++;
			}
		}

		avgrenderframetime /= measuredframes;
		avgrenderframetime += (avgrenderframetime * 0.02f);

		renderframenum++;

		if (renderframenum > 59)
		{
			renderframenum = 0;
		}

		last_was_renderframe = false;
	}

	/* Calculate the average time spend to process a packet
	   frame. Packet frames are mostly dependend on the CPU
	   speed and the network delay. Take the last 60 pure
	   packet frames (frames that are only packet frames ans
	   nothing else) into account and add a security margin
	   of 2%. */
	if (last_was_packetframe && last_was_renderframe)
	{
		int measuredframes = 0;
		static int packetframenum;

		packetframetimes[packetframenum] = msec;

		for (int i = 0; i < 60; i++)
		{
			if (packetframetimes[i] != 0)
			{
				avgpacketframetime += packetframetimes[i];
				measuredframes++;
			}
		}

		avgpacketframetime /= measuredframes;
		avgpacketframetime += (avgpacketframetime * 0.02f);

		packetframenum++;

		if (packetframenum > 59)
		{
			packetframenum = 0;
		}

		last_was_packetframe = false;
	}


	// Calculate timings.
	packetdelta += msec;
	renderdelta += msec;
	clienttimedelta += msec;
	servertimedelta += msec;

	if (!cl_timedemo->value) {
		if (cl_async->value) {
			// Network frames..
			if (packetdelta < ((1000000.0f - avgpacketframetime) / pfps)) {
				packetframe = false;
			}

			// Render frames.
			if (renderdelta < ((1000000.0f - avgrenderframetime) / rfps)) {
				renderframe = false;
			}
		} else {
			// Cap frames at target framerate.
			if (renderdelta < (1000000.0f / rfps)) {
				renderframe = false;
				packetframe = false;
			}
		}
	}
	else if (clienttimedelta < 1000 || servertimedelta < 1000)
	{
		return;
	}


	// Dedicated server terminal console.
	do {
		s = Sys_ConsoleInput();

		if (s) {
			Cbuf_AddText(va("%s\n", s));
		}
	} while (s);

	Cbuf_Execute();


	if (host_speeds->value)
	{
		time_before = Sys_Milliseconds();
	}


	// Run the serverframe.
	if (packetframe) {
		SV_Frame(servertimedelta);
		servertimedelta = 0;
	}


	if (host_speeds->value)
	{
		time_between = Sys_Milliseconds();
	}


	// Run the client frame.
	if (packetframe || renderframe) {
		CL_Frame(packetdelta, renderdelta, clienttimedelta, packetframe, renderframe);
		clienttimedelta = 0;
	}


	if (host_speeds->value)
	{
		int all, sv, gm, cl, rf;

		time_after = Sys_Milliseconds();
		all = time_after - time_before;
		sv = time_between - time_before;
		cl = time_after - time_between;
		gm = time_after_game - time_before_game;
		rf = time_after_ref - time_before_ref;
		sv -= gm;
		cl -= rf;
		Com_Printf("all:%3i sv:%3i gm:%3i cl:%3i rf:%3i\n",
				all, sv, gm, cl, rf);
	}


	// Reset deltas and mark frame.
	if (packetframe) {
		packetdelta = 0;
		last_was_packetframe = true;
	}

	if (renderframe) {
		renderdelta = 0;
		last_was_renderframe = true;
	}
}
#else
void
Qcommon_Frame(int msec)
{
	// For the dedicated server terminal console.
	char *s;

	// Target packetframerate.
	int pfps;

	// Time since last packetframe in microsec.
	static int packetdelta = 1000000;

	// Accumulated time since last server run.
	static int servertimedelta = 0;

	/* A packetframe runs the server and the client,
	   but not the renderer. The minimal interval of
	   packetframes is about 10.000 microsec. If run
	   more often the movement prediction in pmove.c
	   breaks. That's the Q2 variant if the famous
	   125hz bug. */
	qboolean packetframe = true;


	/* In case of ERR_DROP we're jumping here. Don't know
	   if that' really save but it seems to work. So leave
	   it alone. */
	if (setjmp(abortframe))
	{
		return;
	}


	// Timing debug crap. Just for historical reasons.
	if (fixedtime->value)
	{
		msec = (int)fixedtime->value;
	}
	else if (timescale->value)
	{
		msec *= timescale->value;
	}


	// Save global time for network- und input code.
	curtime = Sys_Milliseconds();


	// Target framerate.
	pfps = (int)cl_maxfps->value;


	// Calculate timings.
	packetdelta += msec;
	servertimedelta += msec;


	// Network frame time.
	if (packetdelta < (1000000.0f / pfps)) {
		packetframe = false;
	}


	// Dedicated server terminal console.
	do {
		s = Sys_ConsoleInput();

		if (s) {
			Cbuf_AddText(va("%s\n", s));
		}
	} while (s);

	Cbuf_Execute();


	// Run the serverframe.
	if (packetframe) {
		SV_Frame(servertimedelta);
		servertimedelta = 0;
	}


	// Reset deltas if necessary.
	if (packetframe) {
		packetdelta = 0;
	}
}
#endif

void
Qcommon_Shutdown(void)
{
	Cvar_Fini();
}
